hicommonwealth/commonwealth
# Add Upvote Notifications

## Link to Issue
Closes: #8797 

## Description of Changes
- Added `CommentUpvoted` event
- Updated `ThreadUpvoted` event schema to ensure `comment_id` is not defined
- Updated Rascal (RabbitMQ) configuration to route both upvote event types to the Knock worker (notifications provider)
- Added `ThreadUpvoted` and `CommentUpvoted` event handlers in the notifications policy on the Knock worker
- Added a migration which updates the indexes/primary key on `ThreadSubscriptions`, `CommentSubscriptions`, and `CommunityAlerts` tables.
- Created the upvote workflow and notification templates in Knock

## Test Plan
Requires Knock keys in `.env`. Contact me for keys if you don't have them yet.
- Ensure the `ALLOWED_EVENTS` env var includes `CommentUpvoted` and `ThreadUpvoted`
- Start the internal API, message relayer (`pnpm start-message-relayer`), and Knock worker (`pnpm start-knock`)
- As user A create a thread and ensure you are subscribed to it
- As user B upvote the thread
- User A should receive an upvote notification after the batch time expires. The batch time is set to 3 hours. Every time you upvote (within a 24 hr period) the batch will extend by an additional 3 hours. No more than 1 notification per thread or comment will be sent in 24 hrs and no more than 2 upvote notifications per 24 hrs (irrespective of thread/comment).

## Deployment Plan

1. Promote the upvote workflow on Knock to the production environment
2. Update `ALLOWED_EVENTS` to include `CommentUpvoted`

## Other Considerations

-
